bftengine/src/preprocessor/RequestProcessingInfo.cpp

```diff
@@ -10,30 +10,72 @@
 // file.
 
 #include "RequestProcessingInfo.hpp"
-#include "ReplicaConfig.hpp"
 #include "Logger.hpp"
 
 namespace preprocessor {
 
 using namespace std;
 using namespace concord::util;
-using namespace concordUtils;
 
-RequestProcessingInfo::RequestProcessingInfo(uint16_t numOfReplicas, ReqId reqSeqNum)
+uint16_t RequestProcessingInfo::numOfRequiredEqualReplies_ = 0;
+
+RequestProcessingInfo::RequestProcessingInfo(uint16_t numOfReplicas, uint16_t numOfRequiredReplies, ReqId reqSeqNum)
     : numOfReplicas_(numOfReplicas), reqSeqNum_(reqSeqNum) {
-  for (auto i = 0; i < numOfReplicas; i++)
-    // Placeholders for all replicas
-    replicasDataForRequest_.push_back(nullptr);
+  numOfRequiredEqualReplies_ = numOfRequiredReplies;
   LOG_DEBUG(GL, "Created RequestProcessingInfo with reqSeqNum=" << reqSeqNum_ << ", numOfReplicas= " << numOfReplicas_);
 }
 
-void RequestProcessingInfo::saveClientPreProcessRequestMsg(const ClientPreProcessReqMsgSharedPtr& clientPreProcessReq) {
-  clientPreProcessRequestMsg_ = clientPreProcessReq;
+void RequestProcessingInfo::handlePrimaryPreProcessed(PreProcessRequestMsgSharedPtr msg,
+                                                      const char *preProcessResult,
+                                                      uint32_t preProcessResultLen) {
+  numOfReceivedReplies_++;
+  preProcessRequestMsg_ = msg;
+  myPreProcessResult_ = preProcessResult;
+  myPreProcessResultLen_ = preProcessResultLen;
+  myPreProcessResultHash_ = convertToArray(SHA3_256().digest(myPreProcessResult_, myPreProcessResultLen_).data());
+  preProcessingResultHashes_[myPreProcessResultHash_]++;
+}
+
+void RequestProcessingInfo::handlePreProcessReplyMsg(PreProcessReplyMsgSharedPtr preProcessReplyMsg) {
+  numOfReceivedReplies_++;
+  preProcessingResultHashes_[convertToArray(preProcessReplyMsg->resultsHash())]++;  // Count equal hashes
+}
+
+HashArray RequestProcessingInfo::convertToArray(const uint8_t resultsHash[SHA3_256::SIZE_IN_BYTES]) {
+  HashArray hashArray;
+  for (uint64_t i = 0; i < SHA3_256::SIZE_IN_BYTES; i++) hashArray[i] = resultsHash[i];
+  return hashArray;
 }
 
-void RequestProcessingInfo::savePreProcessResult(const Sliver& preProcessResult, uint32_t preProcessResultLen) {
-  myPreProcessResult_ = preProcessResult.subsliver(0, preProcessResultLen);
-  myPreProcessResultHash_ = SHA3_256().digest(myPreProcessResult_.data(), myPreProcessResult_.length());
+PreProcessingResult RequestProcessingInfo::getPreProcessingConsensusResult() const {
+  if (numOfReceivedReplies_ < numOfRequiredEqualReplies_) return CONTINUE;
```

Comment:
I like the structure of this, and how it returns the enum. It makes it nice and easy to test, as you did below.

Reply:
I prefer to have a raw buffer sent via a network to be sure its internal structure does not change.